### ISSUE TITLE
liftEffect and ordering in orr might lead to "halting" children

### ISSUE THREAD

**Issue by pkamenarsky** [purescript-concur/purescript-concur-react#16]
Here's a minimal test case (as minimal as I could get it down to) running on `v0.3.6`. The idea is to have a looping widget until a button is pressed (think a query form with some internal state) and then return that internal state to the parent widget.

So `effWidget` shows a `div` waiting on a mouse down event, performs a side effect, and then gets called again with a new `b`, in this case just showing a `div` without any attached events.

If the user clicks the "Query" button then `query` returns with some value, which is then shown in `queryTest`. The problem is that once the user clicks on the `div` (which is a big rectangle) with the attached mouse down event the rectangle disappears, and pressing the "Query" button doesn't do anything.

Removing the side effect or swapping the children in `query` (which just calls the `orr` combinator afaics) fixes the problem, so I'm guessing this is some edge case interaction between `Aff` and the evaluation of parallel branches (as far as I understand) in the `Widget` monad. Do you have some other insight that might help?

```
effWidget :: Boolean -> Widget HTML Boolean
effWidget b = do
  _ <- if b
     then D.div [ P.style style ] []
     else do
       _ <- D.div [ P.onMouseDown, P.style style ] []
       liftEffect $ log "Effect"  -- if we remove this line everything works

  pure (not b)
  where
    style =
      { position: "absolute"
      , backgroundColor: "#777"
      , top: "100px"
      , left: "100px"
      , width: "300px"
      , height: "300px"
      }

query :: Widget HTML Boolean
query = go false
  where
    go st = do
      event <- D.div
        []
        [ Left  <$> effWidget st
        , Right <$> D.button [ P.onClick ] [ D.text "Query" ] -- if this goes first everything works
        ]
      case event of
        Left st' -> go st'
        Right _  -> pure st

queryTest :: ∀ a. Widget HTML a
queryTest = do
  a <- query
  D.text (show a)

main :: Effect Unit
main = runWidgetInDom "root" queryTest
```

**Comment by ajnsit** [comment]
Thanks for the bug report! I'll take a look.

**Comment by ajnsit** [comment]
@pkamenarsky I checked in a fix for this. I'm still not convinced that the bug is totally gone, but your example works now.

Basically what was happening was that widgets can perform effects before returning a "continuation" in the free monad. The continuation can then itself perform effects and so on... Until it either returns a result, or a continuation with a `View` and event handlers in `Aff`. The case when a widget returns a view was being handled correctly, but not when the widget is completely effects followed by a result, as in the case of `liftEffect`.

Because of code architecture issues, when effects eventually give way to just returning a result, I was simply wrapping that result in `pure` and returning that as the continuation. Then the system would again try to get a result out of that widget, find the result, rewrap that in a widget and so on. ad infinitum. This caused the effectful widget to never actually return a value and, since it has no UI, effectively disappearing and halting.

Now for the fix, I added a special case to return an `Either result continuation` instead of always returning a continuation. However again due to code architecture issues, I still did have to wrap the `Left result` into a dummy continuation, but just a bit later than before. This seems to have made a crucial difference and fixed the bug. I don't know if it still might cause problems elsewhere, but I have left a comment in the code to force me to take a look at it later.

I have closed this issue, but please feel free to reopen this if needed.

**Comment by ajnsit** [comment]
I released v0.3.7 with this bugfix.

**Comment by pkamenarsky** [comment]
Great, works now! I'll try to dig into the internals somewhen soon and see if there's a way to simplify this a bit if possible.
